From: Suraj Jitindar Singh <surajjs@amazon.com>
Date: Fri, 14 Jul 2023 16:11:40 -0700
Subject: x86/bugs: Allow IBPB to be enabled as a mitigation for RAS Poisoning

Allow for IBPB to be enabled via 'retbleed=ibpb' as a mitigation for RAS
Poisoning, even when the CPU is not affected by RetBleed.

If any other mitigation is selected on the command line when the CPU isn't
affected by retbleed then ignore the selection and don't enable a
mitigation.

Also when the CPU is not affected by retbleed don't default to IBPB. That
is require `retbleed=ibpb` explicitly to enable the mitigation for RAS
poisoning due to the performance impact.

Signed-off-by: Suraj Jitindar Singh <surajjs@amazon.com>
---
 arch/x86/include/asm/cpufeatures.h |  1 +
 arch/x86/kernel/cpu/bugs.c         | 46 +++++++++++++++++++++++++-----
 arch/x86/kernel/cpu/common.c       |  8 +++++-
 3 files changed, 47 insertions(+), 8 deletions(-)

diff --git a/arch/x86/include/asm/cpufeatures.h b/arch/x86/include/asm/cpufeatures.h
--- a/arch/x86/include/asm/cpufeatures.h
+++ b/arch/x86/include/asm/cpufeatures.h
@@ -464,5 +464,6 @@
 #define X86_BUG_RETBLEED		X86_BUG(27) /* CPU is affected by RETBleed */
 #define X86_BUG_EIBRS_PBRSB		X86_BUG(28) /* EIBRS is vulnerable to Post Barrier RSB Predictions */
 #define X86_BUG_SMT_RSB			X86_BUG(29) /* CPU is vulnerable to Cross-Thread Return Address Predictions */
+#define X86_BUG_RAS_POISONING		X86_BUG(30) /* CPU is affected by RAS poisoning */
 
 #endif /* _ASM_X86_CPUFEATURES_H */
diff --git a/arch/x86/kernel/cpu/bugs.c b/arch/x86/kernel/cpu/bugs.c
--- a/arch/x86/kernel/cpu/bugs.c
+++ b/arch/x86/kernel/cpu/bugs.c
@@ -854,12 +854,15 @@ early_param("retbleed", retbleed_parse_cmdline);
 
 #define RETBLEED_UNTRAIN_MSG "WARNING: BTB untrained return thunk mitigation is only effective on AMD/Hygon!\n"
 #define RETBLEED_INTEL_MSG "WARNING: Spectre v2 mitigation leaves CPU vulnerable to RETBleed attacks, data leaks possible!\n"
+#define RETBLEED_IBPB_MSG "WARNING: not automatically selecting RETBLEED mitigation as not affected. Use 'retbleed=ibpb' to explicitly enable\n"
 
 static void __init retbleed_select_mitigation(void)
 {
 	bool mitigate_smt = false;
 
-	if (!boot_cpu_has_bug(X86_BUG_RETBLEED) || cpu_mitigations_off())
+	if ((!boot_cpu_has_bug(X86_BUG_RETBLEED) &&
+	     !boot_cpu_has_bug(X86_BUG_RAS_POISONING)) ||
+	    cpu_mitigations_off())
 		return;
 
 	switch (retbleed_cmd) {
@@ -867,6 +870,8 @@ static void __init retbleed_select_mitigation(void)
 		return;
 
 	case RETBLEED_CMD_UNRET:
+		if (!boot_cpu_has_bug(X86_BUG_RETBLEED))
+			break;
 		if (IS_ENABLED(CONFIG_CPU_UNRET_ENTRY)) {
 			retbleed_mitigation = RETBLEED_MITIGATION_UNRET;
 		} else {
@@ -892,6 +897,10 @@ static void __init retbleed_select_mitigation(void)
 	default:
 		if (boot_cpu_data.x86_vendor == X86_VENDOR_AMD ||
 		    boot_cpu_data.x86_vendor == X86_VENDOR_HYGON) {
+			if (!boot_cpu_has_bug(X86_BUG_RETBLEED)) {
+				pr_err(RETBLEED_IBPB_MSG);
+				break;
+			}
 			if (IS_ENABLED(CONFIG_CPU_UNRET_ENTRY))
 				retbleed_mitigation = RETBLEED_MITIGATION_UNRET;
 			else if (IS_ENABLED(CONFIG_CPU_IBPB_ENTRY) && boot_cpu_has(X86_FEATURE_IBPB))
@@ -951,7 +960,8 @@ static void __init retbleed_select_mitigation(void)
 		}
 	}
 
-	pr_info("%s\n", retbleed_strings[retbleed_mitigation]);
+	if (boot_cpu_has_bug(X86_BUG_RETBLEED))
+		pr_info("%s\n", retbleed_strings[retbleed_mitigation]);
 }
 
 #undef pr_fmt
@@ -2386,6 +2396,18 @@ static ssize_t srbds_show_state(char *buf)
 	return sprintf(buf, "%s\n", srbds_strings[srbds_mitigation]);
 }
 
+static char *ras_poisoning_state(void)
+{
+	if (boot_cpu_has_bug(X86_BUG_RAS_POISONING)) {
+		if (retbleed_mitigation == RETBLEED_MITIGATION_IBPB)
+			return ", RAS-Poisoning: IBPB";
+		else
+			return ", RAS-Poisoning: Vulnerable";
+	} else {
+		return "";
+	}
+}
+
 static ssize_t retbleed_show_state(char *buf)
 {
 	if (retbleed_mitigation == RETBLEED_MITIGATION_UNRET ||
@@ -2394,15 +2416,20 @@ static ssize_t retbleed_show_state(char *buf)
 		boot_cpu_data.x86_vendor != X86_VENDOR_HYGON)
 		    return sprintf(buf, "Vulnerable: untrained return thunk / IBPB on non-AMD based uarch\n");
 
-	    return sprintf(buf, "%s; SMT %s\n",
-			   retbleed_strings[retbleed_mitigation],
+	    return sprintf(buf, "%s; SMT %s%s\n",
+			   boot_cpu_has_bug(X86_BUG_RETBLEED) ?
+			   retbleed_strings[retbleed_mitigation] :
+			   "Not affected",
 			   !sched_smt_active() ? "disabled" :
 			   spectre_v2_user_stibp == SPECTRE_V2_USER_STRICT ||
 			   spectre_v2_user_stibp == SPECTRE_V2_USER_STRICT_PREFERRED ?
-			   "enabled with STIBP protection" : "vulnerable");
+			   "enabled with STIBP protection" : "vulnerable",
+			   ras_poisoning_state());
 	}
 
-	return sprintf(buf, "%s\n", retbleed_strings[retbleed_mitigation]);
+	return sprintf(buf, "%s%s\n", boot_cpu_has_bug(X86_BUG_RETBLEED) ?
+				      retbleed_strings[retbleed_mitigation] :
+				      "Not affected", ras_poisoning_state());
 }
 
 static ssize_t cpu_show_common(struct device *dev, struct device_attribute *attr,
@@ -2452,6 +2479,8 @@ static ssize_t cpu_show_common(struct device *dev, struct device_attribute *attr
 		return mmio_stale_data_show_state(buf);
 
 	case X86_BUG_RETBLEED:
+		/* fallthrough */
+	case X86_BUG_RAS_POISONING:
 		return retbleed_show_state(buf);
 
 	default:
@@ -2516,6 +2545,9 @@ ssize_t cpu_show_mmio_stale_data(struct device *dev, struct device_attribute *at
 
 ssize_t cpu_show_retbleed(struct device *dev, struct device_attribute *attr, char *buf)
 {
-	return cpu_show_common(dev, attr, buf, X86_BUG_RETBLEED);
+	if (boot_cpu_has_bug(X86_BUG_RAS_POISONING))
+		return cpu_show_common(dev, attr, buf, X86_BUG_RAS_POISONING);
+	else
+		return cpu_show_common(dev, attr, buf, X86_BUG_RETBLEED);
 }
 #endif
diff --git a/arch/x86/kernel/cpu/common.c b/arch/x86/kernel/cpu/common.c
--- a/arch/x86/kernel/cpu/common.c
+++ b/arch/x86/kernel/cpu/common.c
@@ -1237,6 +1237,8 @@ static const __initconst struct x86_cpu_id cpu_vuln_whitelist[] = {
 #define RETBLEED	BIT(3)
 /* CPU is affected by SMT (cross-thread) return predictions */
 #define SMT_RSB		BIT(4)
+/* CPU is affected by RAS Poisoning, allowing influencing of return address prediction */
+#define RAS_POISONING	BIT(5)
 
 static const struct x86_cpu_id cpu_vuln_blacklist[] __initconst = {
 	VULNBL_INTEL_STEPPINGS(IVYBRIDGE,	X86_STEPPING_ANY,		SRBDS),
@@ -1268,8 +1270,9 @@ static const struct x86_cpu_id cpu_vuln_blacklist[] __initconst = {
 
 	VULNBL_AMD(0x15, RETBLEED),
 	VULNBL_AMD(0x16, RETBLEED),
-	VULNBL_AMD(0x17, RETBLEED | SMT_RSB),
+	VULNBL_AMD(0x17, RETBLEED | SMT_RSB | RAS_POISONING),
 	VULNBL_HYGON(0x18, RETBLEED | SMT_RSB),
+	VULNBL_AMD(0x19, RAS_POISONING),
 	{}
 };
 
@@ -1390,6 +1393,9 @@ static void __init cpu_set_bug_bits(struct cpuinfo_x86 *c)
 	if (cpu_matches(cpu_vuln_blacklist, SMT_RSB))
 		setup_force_cpu_bug(X86_BUG_SMT_RSB);
 
+	if (cpu_matches(cpu_vuln_blacklist, RAS_POISONING))
+		setup_force_cpu_bug(X86_BUG_RAS_POISONING);
+
 	if (cpu_matches(cpu_vuln_whitelist, NO_MELTDOWN))
 		return;
 

